### ISSUE TITLE
Model loses all event listeners on setting new event dispatcher

### ISSUE THREAD

**Issue by HolgerW1** [laravel/framework#8399]
All existing event listeners for a previously set dispatcher on the `Model` class are lost if a new event dispatcher is set on the Eloquent `Model`.

If the loss is intended then there should be at least one last event fired (sth like `unsubscribed`) or a static method on the class called to inform all listeners of the previous dispatcher about the change.

However imho setting a new dispatcher on the `Model` class should take over all previously set listeners, sth. like:
```
	public static function setEventDispatcher(Dispatcher $dispatcher)
	{
		if (static::$dispatcher && (static::$dispatcher != $dispatcher))
		{
			if ($dispatcher)
			{
				// copy the Model listeners to the new dispatcher
			}
			// or inform the model subclasses about the dispatcher change
		}
		
		static::$dispatcher = $dispatcher;
	}
```

**Comment by HolgerW1** [comment]
Unit-test demonstrating the problem:
```
	public function testDispatcherChange()
	{
		$d1 = new Dispatcher;
		Model::setEventDispatcher($d1);
		
		$model1 = new ModelStub;
		$model1->exists = true;
		$model2 = new ModelStub;
		$model2->exists = true;
		
		$marker = false;
		
		ModelStub::deleting(function($m) use (&$marker) {
			$marker = true;
			return false;
		});
		
		$model1->delete();
		$this->assertTrue($marker);
		$marker = false;
		
		$d2 = new Dispatcher;
		Model::setEventDispatcher($d2);
		$model2->delete();
		$this->assertTrue($marker);
	}
```

The second assertion will always fail because the event handler is gone after setting the second dispatcher.

**Comment by JosephSilber** [comment]
Events are registered with the dispatcher. I don't see how the current behavior would *not* be the expected behavior.

**Comment by HolgerW1** [comment]
I disagree. Model classes and the app should be able to assume that their event listeners don't get silently removed in any case. It should be safe to assume that event listeners get called once they have been registered.
With the current implementation the caller who has set the event listener will never know if the listeners he has set got removed.

Btw. this would have fixed the phpunit problem described in #1181 for good as well.

Anyways, doesn't seem you want to fix this since the ticket already got closed. No point in discussing this any further.

**Comment by GrahamCampbell** [comment]
> silently removed in any case.

No, that's not how php works.
